### ISSUE TITLE
[Logic Bug] A Sepcial SQL Brings  Errors ( SELECT * FROM <table> WHERE <column> BETWEEN <column> AND (CAST(<decimal> AS MONEY)) )

### ISSUE THREAD

**Issue by qwebug** [heterodb/pg-strom#649]
#### Describe:

&#x20;Using SELECT \* FROM \<table> WHERE \<column> BETWEEN \<column> AND (CAST(\<decimal> AS MONEY))  and VACUUM FREEZE ANALYZE \<table>(\<column>) bring different results, when turn on and off pg\_strom.enabled.

#### ***Sql:***

```sql
CREATE TABLE t0(c0 money);
CREATE TABLE t1(LIKE t0);
VACUUM FREEZE ANALYZE t1(c0);
INSERT INTO t1(c0) VALUES((0.1)::MONEY);
INSERT INTO t0(c0) VALUES((0.1)::MONEY);
CREATE SCHEMA extensions;
CREATE EXTENSION pg_strom WITH SCHEMA extensions;
set pg_strom.enabled=on;
SELECT * FROM t1, t0 WHERE (t0.c0) BETWEEN (t1.c0) AND (CAST(0.8 AS MONEY));
```

#### ***Result:***

```bash
 c0 | c0 
----+----
(0 rows)
```

#### ***Sql:***

```sql
set pg_strom.enabled=off;
SELECT * FROM t1, t0 WHERE (t0.c0) BETWEEN (t1.c0) AND (CAST(0.8 AS MONEY));
```

#### ***Result:***

```bash
  c0   |  c0   
-------+-------
 $0.10 | $0.10
(1 row)
```

### Environment:

Pg-strom Version: commit b959fef3547bb9050d8159bb2d8cd4aa4dc918d5

PostgreSQL Version: 15.3 on x86\_64-pc-linux-gnu, compiled by gcc (GCC) 8.5.0 20210514 (Red Hat 8.5.0-18), 64-bit

CUDA Version: 12.0.1

NVIDIA Driver Version: 535.86.05

**Comment by kaigai** [comment]
This query also uses numeric --> money cast, so `1f9fad1749fe22bce33d7a3a30b85ec550081b41` should fix it.

I could not reproduce the problem at the latest commit.

**Comment by qwebug** [comment]
This problem was not fixed at latest commit fb87671957502820db18d679be48b67418070a2e.
It still brings different results, when turn on and off pg_strom.enabled.

**Comment by kaigai** [comment]
Sorry, it was enbugged by the lazy validation of numeric type (this is a feature to reduce GPU cycles when numeric values are not actually referenced in the device code).

Fixed by `b2868103c807f9ed23df6c5e1ed55441f77a1fc7`
